torch/csrc/autograd/profiler.cpp

```diff
@@ -202,31 +203,41 @@ struct ProfilerThreadLocalState
   }
 
   void pushRange(
-      const at::StringView& name,
+      const at::RecordFunction& fn,
       const char* msg = "",
-      int64_t sequence_nr = -1,
-      std::vector<std::vector<int64_t>>&& shapes = {},
-      at::RecordFunctionHandle handle = 0) {
+      std::vector<std::vector<int64_t>>&& shapes = {}) {
     if (config_.state == ProfilerState::Disabled) {
       return;
     }
     if (config_.state == ProfilerState::NVTX) {
       cuda_stubs->nvtxRangePushA(getNvtxStr(
-          name, msg, sequence_nr, shapes).c_str());
+          fn.name(), msg, fn.seqNr(), shapes).c_str());
     } else {
       Event evt(EventKind::PushRange,
-          name,
+          fn.name(),
           at::RecordFunction::currentThreadId(),
           config_.state == ProfilerState::CUDA,
-          handle,
+          fn.handle(),
           std::move(shapes),
           at::RecordFunction::getDefaultNodeId());
-      evt.setSequenceNr(sequence_nr);
+      evt.setSequenceNr(fn.seqNr());
+      evt.setFwdThreadId(fn.forwardThreadId());
+      evt.setScope((uint8_t)fn.scope());
+
+      // backward nodes source range corresponds to the forward node
+      if (config_.with_source && fn.scope() != at::RecordScope::BACKWARD_FUNCTION) {
+        auto cs = jit::currentCallstack();
```

Comment:
I guess if we want to be even fancier - we could unroll the C++ callback (c10::Error has the code for that), then if we encounter any Torchscript in there - append the TS callback and if any python - python callback. That'd make this profiler work even with C++ frontend or with non-trivial custom ops. Downside - it might be spammy with various PT internals from the dispatcher.

I'm not insisting we do it right away, but maybe leave a comment.